### DISCUSSION TITLE
Allow referring to selected formula in other parts of query without repeating formula

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#4554] [Q&A]
**Is your feature request related to a problem? Please describe.**
I want to be able refer to my selected formula column alias elsewhere in the query (for example in orderBy) instead of repeating the formula subquery (which is usually slow)

**Describe the solution you'd like**
Maybe it can detect if the formula column has been selected and then reuse its alias instead of repeating it. Or maybe we can use the  `raw()` and `as()`  helpers from v6 to define an alias in the formula defintion.

**Describe alternatives you've considered**
I've tried to use the query builder and select everything myself but I ran into many unrelated problems with that.

**Additional context**
See discussion #4540 for my specific use case.

**Comment by B4nan** [comment]
IIRC the problem here is that in some parts you need to repeat it, e.g. in where query.

**Comment by squivix** [comment]
> IIRC the problem here is that in some parts you need to repeat it, e.g. in where query.

I don't understand what you mean. Could you give an example?

**Comment by B4nan** [comment]
You can't refer to a select clause from inside where clause, as the select clause is not evaluated yet.

**Comment by squivix** [comment]
I see.

Well we can at least support it for the clauses that evaluate after the selecting is done, such as ORDER BY. That's my main problem really. A subquery in ORDER BY is very slow and it's a shame to repeat it when it's already been selected and aliased.

Alternatively allow setting an alias for the formula explicitly maybe using  v6's `raw` and `alias` helpers. I don't know.

**Comment by B4nan** [comment]
I don't understand how is this related to aliasing, the formula property already has a name and is aliased based on it.

**Comment by squivix** [comment]
> I don't understand how is this related to aliasing, the formula property already has a name and is aliased based on it.

You're right. I was confused. The alias is unrelated.

Anyway I think the enhancement would make formulas better.

**Comment by B4nan** [comment]
Yeah, agreed it would be better, and it should be relatively simple.

**Comment by B4nan** [comment]
Btw just found this, where it says that `order by` is actually the only clause where you can refer to selected items:

https://stackoverflow.com/questions/11182339/reference-alias-calculated-in-select-in-where-clause

I thought this is also possible with `having` clause, need to check that.

**Comment by B4nan** [comment]
That post also mentions that the engine can be capable of caching the expression (it mentions sql server), so would be good if you confirm this actually makes a difference perf-wise.

**Comment by squivix** [comment]
One complication might be the pagination that uses SELECT IN in `wrapPaginatedQuery`. 

It seems like the formula isn't selected in the subquery but it tries to order by it anyway.

I don't know if this is intended behavior but if we replace it with an alias in the order by of the subquery without selecting then it will fail.

**Comment by squivix** [comment]
> so would be good if you confirm this actually makes a difference perf-wise.

It makes a huge difference for me. I have something like 20,000 records in postgres db. Selecting, ordering by formula column, limit 25,the query takes 1 minute.

If I change the query to use ORDER BY formula_column instead of ORDER BY (formula subquery) it takes around 200ms.

This is probably because the formula involves a subquery, but still.

**Comment by B4nan** [comment]
> This is probably because the formula involves a subquery, but still.

Please compare the query execution directly, and try to have it as generated, and then only change the proposed things, so only replace the subquery in order by clause to use the selected alias (if its not selected, add it to the select clause).

**Comment by B4nan** [comment]
Also, it would be good if you provide a complete repro for the perf issue, so I know what changes actually make a difference. e.g. it could be something else with the pagination.

**Comment by squivix** [comment]
After spending some time creating I repro I realized the performance issue was not this. I misjudged.

Specifically I ran the following queries:

```sql
SELECT ea.*,
       (SELECT COUNT(entity_b_id) FROM entity_a_entity_b_pivot WHERE entity_a_id = ea.id) AS entity_bs_count,
       entity_b.*
FROM entity_a ea
         LEFT JOIN entity_a_entity_b_pivot ON ea.id = entity_a_entity_b_pivot.entity_a_id
         LEFT JOIN entity_b ON entity_a_entity_b_pivot.entity_b_id = entity_b.id
GROUP BY ea.id, entity_b.id
ORDER BY (SELECT COUNT(entity_b_id) FROM entity_a_entity_b_pivot WHERE entity_a_id = ea.id) desc, ea.id
LIMIT 25;
```
And:

```sql
SELECT ea.*,
       (SELECT COUNT(entity_b_id) FROM entity_a_entity_b_pivot WHERE entity_a_id = ea.id) AS entity_bs_count,
       entity_b.*
FROM entity_a ea
         LEFT JOIN entity_a_entity_b_pivot ON ea.id = entity_a_entity_b_pivot.entity_a_id
         LEFT JOIN entity_b ON entity_a_entity_b_pivot.entity_b_id = entity_b.id
GROUP BY ea.id, entity_b.id
ORDER BY entity_bs_count desc, ea.id
LIMIT 25;
```

And there was no significant difference.

I found out the performance issue I was seeing was due to something unrelated: The formula being used in the ORDER BY of the inner subquery used for pagination.

I'm still getting to the root of it, but the point is there was no performance difference and I should have tried to repro before reporting. Sorry about that.

I guess the only reason to do this now is for getting a more elegant, better looking query given that the DB engine seems to already optimize the performance. At least postgres.

**Comment by B4nan** [comment]
> I guess the only reason to do this now is for getting a more elegant, better looking query given that the DB engine seems to already optimize the performance. At least postgres.

Yeah, in other words, not really worth it :] 

Will wait for the complete repro then, but it sounds a bit like you simply fall into creating temp tables (because you want to order by a computed expression), and it's irrelevant whether it's being computed in select or order by clause. That itself sounds quite like a wontfix, as the pagination mechanism cannot work only with joins (the whole point of it is to get around issues with joins producing cartesian product). 

I think you should store the count directly on the root entity and maintain the value (e.g. via some hooks, or update it periodically). Right now you basically do the cartesian product twice (as you have two to-many joins), so you end up with a lot of records in memory. Also, think about using select-in strategy, that's often much more performant than a single complex query with many joins.

**Comment by squivix** [comment]
Actually after a bit of investigation, I found out my problem could be solved with better indexes.

I was indexing the joint fk columns of the pivot table but then counting in the subquery based on a filter on one of them. And that column was the second in the composite index. So the count itself was using seq scan. Reversing the order of the columns in the index fixed the slowness. I'm also considering adding a single index on both columns. 

I know this may be taking the easy way out and maybe I should try to write better queries that don't need to order based on a subquery in the first place but it works for me.

I'll try the select in as well see if that makes a difference but my problem is solved and it was unrelated. 

Anyway you're right that this is not worth it, so feel free to close this and thanks for your help.
